### PR TITLE
Ibc switch

### PR DESCRIPTION
**Pull request by eshelB**
Still TODO in this PR: integrate the middleware on all stacks.
Test (although that will be done in secretjs)
Fix some todos that I left on the code

### CODE REVIEW

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -85,8 +87,10 @@ type SecretAppKeepers struct {
 	IbcKeeper        *ibckeeper.Keeper // IBC Keeper must be a pointer in the app, so we can SetRouter on it correctly
 	TransferKeeper   ibctransferkeeper.Keeper
 
-	IbcFeeKeeper        ibcfeekeeper.Keeper
-	PacketForwardKeeper *ibcpacketforwardkeeper.Keeper
+	IbcFeeKeeper         ibcfeekeeper.Keeper
+	PacketForwardKeeper  *ibcpacketforwardkeeper.Keeper
+	IbcSwitchICS4Wrapper *ibcswitch.ICS4Wrapper
+	TransferStack        *ibcswitch.IBCModule
```
Comment (Cashmaney) [inline app/keepers/keepers.go:93]: Why is this here?

**Review by Cashmaney — CHANGES_REQUESTED** [review]

File: app/keepers/keepers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (eshelB) [reply app/keepers/keepers.go:98]: Removing, was like that in osmosis

File: x/compute/internal/keeper/test_common.go [hunk 2]
```diff
@@ -5,14 +5,15 @@ import (
...
 	"os"
 	"path/filepath"
 	"testing"
 	"time"
 
+	cosmwasm "github.com/scrtlabs/SecretNetwork/go-cosmwasm/types"
+
+	v010cosmwasm "github.com/scrtlabs/SecretNetwork/go-cosmwasm/types/v010"
+
 	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
```
[the remark this answers is not in the document]
Comment (eshelB) [reply x/compute/internal/keeper/test_common.go:17]: Ok, So disregard the failed tests?

File: x/ibc-switch/ics4wrapper.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/scrtlabs/SecretNetwork/x/ibc-switch/types"
+)
+
+var (
+	_ porttypes.Middleware  = &IBCModule{}
+	_ porttypes.ICS4Wrapper = &ICS4Wrapper{}
+)
+
+type ICS4Wrapper struct {
```
[the remark this answers is not in the document]
Comment (eshelB) [reply x/ibc-switch/ics4wrapper.go:20]: Yes, can call it channel. That's the name osmosis uses so I went along with it. renaming

File: app/keepers/keepers.go [hunk 4]
```diff
@@ -295,24 +310,21 @@ func (ak *SecretAppKeepers) InitCustomKeepers(
...
+		ak.GetSubspace("ibc-switch"),
 	)
+	ak.IbcSwitchICS4Wrapper = &ibcSwitchICS4Wrapper
 
 	ak.IbcFeeKeeper = ibcfeekeeper.NewKeeper(
 		appCodec,
 		ak.keys[ibcfeetypes.StoreKey],
 		ak.GetSubspace(ibcfeetypes.ModuleName), // this isn't even used in the keeper but is required?
-		ak.IbcKeeper.ChannelKeeper,             // replaced with IBC middleware
+		ak.IbcSwitchICS4Wrapper,                // integrate ibc-switch with every app that uses ibc fees middleware
```
[the remark this answers is not in the document]
Comment (eshelB) [reply app/keepers/keepers.go]: Seems that some middlewares call their keeper's Send() instead of their underlying app's Send(). This is the case for ibcFee.

File: app/keepers/keepers.go [hunk 5]
```diff
@@ -350,7 +364,10 @@ func (ak *SecretAppKeepers) InitCustomKeepers(
 		appCodec,
 		ak.keys[ibctransfertypes.StoreKey],
 		ak.GetSubspace(ibctransfertypes.ModuleName),
-		ak.PacketForwardKeeper,
+		// todo: verify the following: the transfer keeper does not need to know about packet forward keeper, because
+		//  we don't want to go through forward module if the packets originated in this chain.
+		// todo: verify the following: we want fees for the transfer app (it previously didn't have)
+		ak.IbcFeeKeeper, // integrate fee channel with transfer
```
[the remark this answers is not in the document]
Comment (eshelB) [reply app/keepers/keepers.go]: Osmosis doesn't bother with PacketForward on the Send direction, since PacketForward can become active only on receive direction. But I returned it to be present in both directions of the stack after I merged Assaf's Ibc-hooks, for simplicity

File: app/keepers/keepers.go [hunk 6]
```diff
@@ -323,7 +335,8 @@ func (ak *SecretAppKeepers) InitCustomKeepers(
 		appCodec,
 		ak.keys[icacontrollertypes.StoreKey],
 		ak.GetSubspace(icacontrollertypes.SubModuleName),
-		ak.IbcFeeKeeper, // integrate fee keeper with ica
+		// todo: how can this work if IbcFeeKeeper does not implement ics4Wrapper?? Juno seems to have a bug
```
[the remark this answers is not in the document]
Comment (eshelB) [reply app/keepers/keepers.go]: Removed it. It was a work-in-progress. (I wrote on the description of this PR that I still have some todos to fix).

File: app/keepers/keepers.go [hunk 7]
```diff
@@ -373,7 +401,12 @@ func (ak *SecretAppKeepers) InitCustomKeepers(
 		ibcpacketforwardkeeper.DefaultRefundTransferPacketTimeoutTimestamp,
 	)
 	transferStack = ibcfee.NewIBCMiddleware(transferStack, ak.IbcFeeKeeper)
+	// todo: this is ugly since the IBCModule interface on the switch module is implemented with pointers, try without pointers instead
+	var stackWithSwitch ibcswitch.IBCModule
```
[the remark this answers is not in the document]
Comment (eshelB) [reply app/keepers/keepers.go:398]: fixed, it was an ugly solution because of the implemetation of the interface with `func (m *IBCModule) ...` (with a pointer) instead of without a pointer.

File: x/ibc-switch/types/keys.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+var RouterKey = strings.ReplaceAll(ModuleName, "-", "")
+
+var (
+	KeySwitchStatus  = []byte("switchstatus")
+	KeyPauserAddress = []byte("pauseraddress")
+)
+
+const (
+	IbcSwitchStatusOff string = "off"
+	IbcSwitchStatusOn  string = "on"
```
[the remark this answers is not in the document]
Comment (eshelB) [reply x/ibc-switch/types/keys.go:20]: I thought this one would be more direct since you can't go wrong if it's "off", but if it was 0 you wouldn't know if the switch lets packets through or doesn't. But maybe you're right. Although maybe not essential since I would have to change all the messages in secretjs?
